File: board/move_generator.py
#move_generator.py

from board.move import Move


def generate_all_moves(board, color):

    moves = []

    for r in range(8):
        for c in range(8):

            piece = board.board[r][c]

            if piece == "":
                continue

            if color == "white" and piece[0] != "w":
                continue

            if color == "black" and piece[0] != "b":
                continue

            p = piece[1]

            if p == "P":
                pawn_moves(board, r, c, moves)

            elif p == "N":
                knight_moves(board, r, c, moves)

            elif p == "B":
                bishop_moves(board, r, c, moves)

            elif p == "R":
                rook_moves(board, r, c, moves)

            elif p == "Q":
                bishop_moves(board, r, c, moves)
                rook_moves(board, r, c, moves)

            elif p == "K":
                king_moves(board, r, c, moves)

    return moves


def pawn_moves(board, r, c, moves):

    piece = board.board[r][c]

    direction = -1 if piece[0] == "w" else 1
    start_row = 6 if piece[0] == "w" else 1

    if 0 <= r + direction < 8 and board.board[r + direction][c] == "":
        end_row = r + direction

        if end_row == 0 or end_row == 7:
            moves.append(Move((r,c),(end_row,c),piece,"",promotion="Q"))
        else:
            moves.append(Move((r,c),(end_row,c),piece,""))

            if r == start_row and 0 <= r + 2*direction < 8 and board.board[r + 2*direction][c] == "":
                moves.append(Move((r,c),(r + 2*direction,c),piece,""))

    for dc in [-1,1]:

        col = c + dc
        row = r + direction

        if 0 <= col < 8 and 0 <= row < 8:

            target = board.board[row][col]

            if target != "" and target[0] != piece[0]:
                moves.append(Move((r,c),(row,col),piece,target))

            # ✅ EN PASSANT FIX
            if board.en_passant_target == (row, col):

                captured = "bP" if piece[0] == "w" else "wP"

                move = Move((r, c), (row, col), piece, captured)
                move.is_en_passant = True

                if piece[0] == "w":
                    move.en_passant_capture_pos = (row + 1, col)
                else:
                    move.en_passant_capture_pos = (row - 1, col)

                moves.append(move)


def knight_moves(board, r, c, moves):

    piece = board.board[r][c]

    dirs = [
        (-2, -1), (-2, 1),
        (-1, -2), (-1, 2),
        (1, -2), (1, 2),
        (2, -1), (2, 1)
    ]

    for dr, dc in dirs:

        row = r + dr
        col = c + dc

        if 0 <= row < 8 and 0 <= col < 8:

            target = board.board[row][col]

            if target == "" or target[0] != piece[0]:
                moves.append(Move((r, c), (row, col), piece, target))


def rook_moves(board, r, c, moves):
    slide(board, r, c, moves, [(1, 0), (-1, 0), (0, 1), (0, -1)])


def bishop_moves(board, r, c, moves):
    slide(board, r, c, moves, [(1, 1), (1, -1), (-1, 1), (-1, -1)])


def slide(board, r, c, moves, dirs):

    piece = board.board[r][c]

    for dr, dc in dirs:

        row = r + dr
        col = c + dc

        while 0 <= row < 8 and 0 <= col < 8:

            target = board.board[row][col]

            if target == "":
                moves.append(Move((r, c), (row, col), piece, ""))

            else:
                if target[0] != piece[0]:
                    moves.append(Move((r, c), (row, col), piece, target))
                break

            row += dr
            col += dc


# move_generator.py
# replace only the king_moves function

def king_moves(board, r, c, moves):

    piece = board.board[r][c]
    color = "white" if piece[0] == "w" else "black"

    dirs = [
        (-1,-1),(-1,0),(-1,1),
        (0,-1),(0,1),
        (1,-1),(1,0),(1,1)
    ]

    for dr, dc in dirs:

        row = r + dr
        col = c + dc

        if 0 <= row < 8 and 0 <= col < 8:

            target = board.board[row][col]

            if target == "" or target[0] != piece[0]:
                moves.append(Move((r,c),(row,col),piece,target))

    # ✅ CASTLING — all 3 rules enforced
    from board.rule_checker import is_in_check

    # Rule 1 — cannot castle while in check
    if is_in_check(board, color):
        return

    if piece == "wK":

        if board.castling_rights["wK"]:
            if board.board[7][5] == "" and board.board[7][6] == "":

                # Rule 2 — king cannot pass through attacked square
                if not _square_attacked(board, 7, 5, "b") and \
                   not _square_attacked(board, 7, 6, "b"):

                    move = Move((7,4),(7,6),piece,"")
                    move.is_castling = True
                    moves.append(move)

        if board.castling_rights["wQ"]:
            if board.board[7][3] == "" and board.board[7][2] == "" and board.board[7][1] == "":

                # Rule 2 — king cannot pass through attacked square
                if not _square_attacked(board, 7, 3, "b") and \
                   not _square_attacked(board, 7, 2, "b"):

                    move = Move((7,4),(7,2),piece,"")
                    move.is_castling = True
                    moves.append(move)

    elif piece == "bK":

        if board.castling_rights["bK"]:
            if board.board[0][5] == "" and board.board[0][6] == "":

                # Rule 2 — king cannot pass through attacked square
                if not _square_attacked(board, 0, 5, "w") and \
                   not _square_attacked(board, 0, 6, "w"):

                    move = Move((0,4),(0,6),piece,"")
                    move.is_castling = True
                    moves.append(move)

        if board.castling_rights["bQ"]:
            if board.board[0][3] == "" and board.board[0][2] == "" and board.board[0][1] == "":

                # Rule 2 — king cannot pass through attacked square
                if not _square_attacked(board, 0, 3, "w") and \
                   not _square_attacked(board, 0, 2, "w"):

                    move = Move((0,4),(0,2),piece,"")
                    move.is_castling = True
                    moves.append(move)

# move_generator.py
# add at the bottom of the file

def _square_attacked(board, row, col, attacker_color):
    """
    Check if a specific square is attacked by any piece of attacker_color.
    Used for castling validation only.
    """

    # pawn attacks
    if attacker_color == "w":
        pawn_dirs = [(-1, -1), (-1, 1)]
    else:
        pawn_dirs = [(1, -1), (1, 1)]

    for dr, dc in pawn_dirs:
        r, c = row + dr, col + dc
        if 0 <= r < 8 and 0 <= c < 8:
            if board.board[r][c] == attacker_color + "P":
                return True

    # knight attacks
    for dr, dc in [(-2,-1),(-2,1),(-1,-2),(-1,2),(1,-2),(1,2),(2,-1),(2,1)]:
        r, c = row + dr, col + dc
        if 0 <= r < 8 and 0 <= c < 8:
            if board.board[r][c] == attacker_color + "N":
                return True

    # rook / queen (straight lines)
    for dr, dc in [(1,0),(-1,0),(0,1),(0,-1)]:
        r, c = row + dr, col + dc
        while 0 <= r < 8 and 0 <= c < 8:
            piece = board.board[r][c]
            if piece != "":
                if piece[0] == attacker_color and piece[1] in ["R", "Q"]:
                    return True
                break
            r += dr
            c += dc

    # bishop / queen (diagonals)
    for dr, dc in [(1,1),(1,-1),(-1,1),(-1,-1)]:
        r, c = row + dr, col + dc
        while 0 <= r < 8 and 0 <= c < 8:
            piece = board.board[r][c]
            if piece != "":
                if piece[0] == attacker_color and piece[1] in ["B", "Q"]:
                    return True
                break
            r += dr
            c += dc

    # king attacks
    for dr, dc in [(-1,-1),(-1,0),(-1,1),(0,-1),(0,1),(1,-1),(1,0),(1,1)]:
        r, c = row + dr, col + dc
        if 0 <= r < 8 and 0 <= c < 8:
            if board.board[r][c] == attacker_color + "K":
                return True

    return False

File: server-java/src/main/java/com/pace/chess/engine/MoveGenerator.java
package com.pace.chess.engine;

import com.pace.chess.model.Board;
import com.pace.chess.model.Move;
import com.pace.chess.model.Piece;

import java.util.ArrayList;
import java.util.List;

public class MoveGenerator {

    private static final int[] KNIGHT_DIRS = {-33, -31, -18, -14, 14, 18, 31, 33};
    private static final int[] BISHOP_DIRS = {-17, -15, 15, 17};
    private static final int[] ROOK_DIRS   = {-16, -1, 1, 16};
    private static final int[] KING_DIRS   = {-17, -16, -15, -1, 1, 15, 16, 17};

    public static List<Move> generateAllMoves(Board board, int color) {
        List<Move> moves = new ArrayList<>();
        int enemyColor = (color == Piece.WHITE) ? Piece.BLACK : Piece.WHITE;

        // Loop through the 0x88 board
        for (int sq = 0; sq < 128; sq++) {
            // Skip "ghost" squares off the board
            if ((sq & 0x88) != 0) continue;

            int piece = board.board[sq];
            if (piece == Piece.EMPTY || (piece & (Piece.WHITE | Piece.BLACK)) != color) continue;

            int type = piece & 7; // Extract piece type (1-6)

            switch (type) {
                case Piece.PAWN -> pawnMoves(board, sq, color, enemyColor, moves);
                case Piece.KNIGHT -> stepMoves(board, sq, piece, enemyColor, KNIGHT_DIRS, moves);
                case Piece.BISHOP -> slideMoves(board, sq, piece, enemyColor, BISHOP_DIRS, moves);
                case Piece.ROOK -> slideMoves(board, sq, piece, enemyColor, ROOK_DIRS, moves);
                case Piece.QUEEN -> {
                    slideMoves(board, sq, piece, enemyColor, BISHOP_DIRS, moves);
                    slideMoves(board, sq, piece, enemyColor, ROOK_DIRS, moves);
                }
                case Piece.KING -> kingMoves(board, sq, piece, color, enemyColor, moves);
            }
        }
        return moves;
    }

    private static void pawnMoves(Board board, int sq, int color, int enemyColor, List<Move> moves) {
        int dir = (color == Piece.WHITE) ? -16 : 16;
        int startRank = (color == Piece.WHITE) ? 6 : 1;

        int piece = color | Piece.PAWN;
        int currentRank = sq / 16;

        // 1. Forward Push
        int target = sq + dir;
        if ((target & 0x88) == 0 && board.board[target] == Piece.EMPTY) {
            addPawnMove(sq, target, piece, Piece.EMPTY, currentRank == startRank + (dir > 0 ? 5 : -5), moves);
            
            // 2. Double Push
            int doubleTarget = sq + (dir * 2);
            if (currentRank == startRank && board.board[doubleTarget] == Piece.EMPTY) {
                moves.add(new Move(sq, doubleTarget, piece, Piece.EMPTY));
            }
        }

        // 3. Captures
        int[] capDirs = (color == Piece.WHITE) ? new int[]{-17, -15} : new int[]{15, 17};
        for (int cd : capDirs) {
            target = sq + cd;
            if ((target & 0x88) == 0) {
                int targetPiece = board.board[target];
                // Normal Capture
                if (targetPiece != Piece.EMPTY && (targetPiece & enemyColor) != 0) {
                    addPawnMove(sq, target, piece, targetPiece, currentRank == startRank + (dir > 0 ? 5 : -5), moves);
                }
                // En Passant
                else if (board.enPassantTarget != null && target == board.enPassantTarget) {
                    Move m = new Move(sq, target, piece, enemyColor | Piece.PAWN);
                    m.isEnPassant = true;
                    m.enPassantCapturePos = target - dir;
                    moves.add(m);
                }
            }
        }
    }

    private static void addPawnMove(int sq, int target, int piece, int captured, boolean isPromo, List<Move> moves) {
        if (isPromo) {
            moves.add(new Move(sq, target, piece, captured, Piece.QUEEN));
            moves.add(new Move(sq, target, piece, captured, Piece.ROOK));
            moves.add(new Move(sq, target, piece, captured, Piece.BISHOP));
            moves.add(new Move(sq, target, piece, captured, Piece.KNIGHT));
        } else {
            moves.add(new Move(sq, target, piece, captured));
        }
    }

    private static void stepMoves(Board board, int sq, int piece, int enemyColor, int[] dirs, List<Move> moves) {
        for (int d : dirs) {
            int target = sq + d;
            if ((target & 0x88) == 0) {
                int targetPiece = board.board[target];
                if (targetPiece == Piece.EMPTY || (targetPiece & enemyColor) != 0) {
                    moves.add(new Move(sq, target, piece, targetPiece));
                }
            }
        }
    }

    private static void slideMoves(Board board, int sq, int piece, int enemyColor, int[] dirs, List<Move> moves) {
        for (int d : dirs) {
            int target = sq + d;
            while ((target & 0x88) == 0) {
                int targetPiece = board.board[target];
                if (targetPiece == Piece.EMPTY) {
                    moves.add(new Move(sq, target, piece, Piece.EMPTY));
                } else {
                    if ((targetPiece & enemyColor) != 0) {
                        moves.add(new Move(sq, target, piece, targetPiece)); // Capture
                    }
                    break; // Hit a piece, stop sliding
                }
                target += d;
            }
        }
    }

    private static void kingMoves(Board board, int sq, int piece, int color, int enemyColor, List<Move> moves) {
        // Normal moves
        stepMoves(board, sq, piece, enemyColor, KING_DIRS, moves);

        // Castling (Cannot castle out of, through, or into check)
        if (RuleChecker.isInCheck(board, color)) return;

        if (color == Piece.WHITE) {
            if (board.castlingRights.getOrDefault("wK", false) && 
                board.board[sq + 1] == Piece.EMPTY && board.board[sq + 2] == Piece.EMPTY &&
                !RuleChecker.isSquareAttacked(board, sq + 1, enemyColor) && !RuleChecker.isSquareAttacked(board, sq + 2, enemyColor)) {
                Move m = new Move(sq, sq + 2, piece, Piece.EMPTY); m.isCastling = true; moves.add(m);
            }
            if (board.castlingRights.getOrDefault("wQ", false) && 
                board.board[sq - 1] == Piece.EMPTY && board.board[sq - 2] == Piece.EMPTY && board.board[sq - 3] == Piece.EMPTY &&
                !RuleChecker.isSquareAttacked(board, sq - 1, enemyColor) && !RuleChecker.isSquareAttacked(board, sq - 2, enemyColor)) {
                Move m = new Move(sq, sq - 2, piece, Piece.EMPTY); m.isCastling = true; moves.add(m);
            }
        } else {
            if (board.castlingRights.getOrDefault("bK", false) && 
                board.board[sq + 1] == Piece.EMPTY && board.board[sq + 2] == Piece.EMPTY &&
                !RuleChecker.isSquareAttacked(board, sq + 1, enemyColor) && !RuleChecker.isSquareAttacked(board, sq + 2, enemyColor)) {
                Move m = new Move(sq, sq + 2, piece, Piece.EMPTY); m.isCastling = true; moves.add(m);
            }
            if (board.castlingRights.getOrDefault("bQ", false) && 
                board.board[sq - 1] == Piece.EMPTY && board.board[sq - 2] == Piece.EMPTY && board.board[sq - 3] == Piece.EMPTY &&
                !RuleChecker.isSquareAttacked(board, sq - 1, enemyColor) && !RuleChecker.isSquareAttacked(board, sq - 2, enemyColor)) {
                Move m = new Move(sq, sq - 2, piece, Piece.EMPTY); m.isCastling = true; moves.add(m);
            }
        }
    }
}